page/index.md

```diff
@@ -282,4 +302,17 @@ transformations that checks that the field being manipulated is lazy (via
 `islazyfield`) and converts `getproperty` and `setproperty!` to `getfield` and
 `setfield!`.
 
+For immutable structs without type parameters we do an optimization to keep the
+struct mutable and define it as
```

Comment:
This sounds like it would always generate worse code than just making it mutable, so calling it an optimization seems misleading. Perhaps say something like "this will make a larger, slower, fragmented layout in order to make just those fields mutable"?

Reply:
Having those fields non mutable is a feature. It's up to you to make it fast ;).

Reply:
You're semantically requiring it to be slower though, so there's not much the compiler can do to undo that (other than convert it back to a mutable struct, if you're very lucky). You requiring the computer to do more work at runtime.

Reply:
> (other than convert it back to a mutable struct, if you're very lucky)

It can do whatever it wants as long as `setfield` on the fields that are suppose to be immutable, errors. But we are all just beating around the bush that we don't have

```jl
mutable struct Foo
    @const x::Int
    @const y::Int
    z::Int
end
```

Reply:
> It can do whatever it wants as long as setfield on the fields that are suppose to be immutable, errors. 

Because `Ref` (or `Box`) is mutable, it needs its own identity (must be GC'able independently) and so it cannot be allocated inline in `struct Foo` (AFAICT). So you're stuck with n_lazy_fields small allocations instead of one bigger allocation, which was @vtjnash's point.

Reply:
Yes, the current implementation is the only way to get the desired semantics. Whenever there is a better way, I'll switch to it. And I probably won't merge this PR until there is one.